### PR TITLE
feat: add audit-derived draft-gate personas + script resilience (#418)

### PR DESCRIPTION
**Pull request by mfittko**
## Summary

Act on Copilot comment audit findings (#363/#368):
- Derive 5 new draft-gate personas from audit-identified gap categories
- Make audit script resilient to rate limits and interruptible
- Document the audit script with full taxonomy and recommendation mapping

### 1. New draft-gate personas (`.pi/dev-loop/defaults.yaml`)

| Persona | Gate wiring | Source category | Comments |
|---|---|---|---|
| `ci-guard` | **Default** draft gate | ci_guard | 67 |
| `link-check` | Opt-in draft gate | placeholder_404 + broken_paths | 36 |
| `config-drift` | Opt-in draft gate | config_conflicts | 27 |
| `gate-evidence` | Opt-in draft gate | gate_evidence | 13 |
| `no-op` | Opt-in draft gate | no_op_tool_usage | 11 |

### 2. Strengthened existing personas

- `coverage`: added test-name quality, edge-case coverage, malformed-argument tests, negative-case coverage language
- `docs`: added stale command reference language

### 3. Script resilience (`scripts/github/audit-copilot-comments.mjs`)

- Retry with exponential backoff on 403/429 (5 retries, 1s base)
- 401 fails immediately
- `--sleep-ms <ms>` flag between top-level fetches
- `--checkpoint-file <path>` + `--resume` for coarse-grain resume (after-comments / after-prs stages)
- Checkpoint corruption falls back to fresh fetch

### 4. Documentation (`docs/audit-copilot-comments.md`)

- Purpose, usage, output format
- Full 12-category taxonomy table
- Recommendation lens mapping
- Draft-gate persona derivation from audit findings
- Edge cases
- Linked from `docs/index.md`

## Acceptance criteria

- [x] `docs/audit-copilot-comments.md` exists, covers purpose/usage/output/taxonomy, linked from `docs/index.md`
- [x] 5 new draft-gate personas with prompt text in `.pi/dev-loop/defaults.yaml`
- [x] `ci-guard` wired into default draft gate angles, rest opt-in with comments
- [x] Existing `coverage` and `docs` prompts strengthened with audit-derived guidance
- [x] Retry+backoff on 403/429
- [x] Checkpoint save/resume round-trip
- [x] `--sleep-ms` flag
- [x] `npm test` passes (824 + 702 + 10 = 1536, 0 fail)
- [x] `npm run verify` passes (docs links: 61 files, 302 links OK)

## Definition of done

- [x] Docs entry covers purpose/usage/output/taxonomy
- [x] Retry+backoff tested
- [x] Checkpoint/resume tested
- [x] `--sleep-ms` tested
- [x] Strengthened coverage/docs prompts
- [x] `npm run verify` passes

## Non-goals

- Reclassifying the 76% uncategorized comments (separate follow-up)
- Modeling personas from low-frequency categories
- Adding personas to pre-approval gate (draft gate only for now)
- Re-running the full Copilot comment analysis pipeline
- Changing audit script taxonomy, classification logic, or category definitions

### CODE REVIEW

**Comment by mfittko** [comment]
### Gate review: `draft_gate`

**Reviewed head SHA:** `86ed3b8ac61c60a9b12f163f635b46fa843de3e2`
**Verdict:** clean

**Findings summary:** Draft gate review: scope, coverage, correctness, ci-guard — all clean. **scope**: All 5 files directly implement #418 ACs. Only personas, script resilience, tests, docs. No scope creep. **coverage**: 1536 tests pass (824 scripts + 702 core + 10 dev-loop). New tests cover retry (403, 429, max-retries, 401), checkpoint save/resume/from-comments/corrupt, CLI flag parsing for --sleep-ms, --checkpoint-file, --resume. **correctness**: 5 personas (ci-guard, link-check, config-drift, gate-evidence, no-op) wired. ci-guard in default draft gate. coverage + docs prompts strengthened per audit. **ci-guard**: PR is draft, verify passes. CI config unchanged.

**Next action:** Mark PR ready for Copilot review.

File: scripts/github/audit-copilot-comments.mjs [hunk 1]
```diff
@@ -195,7 +212,7 @@ const CATEGORY_DEFINITIONS = [
       /error-contract/i,
       /malformed-argument/i,
       /happy\s+path\s+only/i,
-      /doesn['’]t\s+exercise/i,
+      /doesn['']t\s+exercise/i,
       /should\s+test/i,
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/github/audit-copilot-comments.mjs:216]: All 6 findings fixed in 7e828e1:

1. **curly-apostrophe patterns**: Restored U+2019 matching in incomplete_coverage and misleading_tests regex — changed from /doesn['']t/ back to /doesn['’]t/ (straight + curly right quote).
2. **--sleep-ms without --checkpoint-file**: Sleep now applies between fetches unconditionally, not only when checkpoint-file is set.
3. **checkpoint parent directory**: saveCheckpoint now creates parent directories via mkdir(recursive: true) before writing.
4. **gate-evidence prompt**: Rewrote 'runs on draft gate by default' to 'opt-in for the draft gate (uncomment to add)'.
5. **--checkpoint-file whitespace**: Added non-empty validation (empty after trim → parseError).
6. Added corresponding test for whitespace-only --checkpoint-file rejection.

825 tests pass (scripts), full verify green.

File: scripts/github/audit-copilot-comments.mjs [hunk 2]
```diff
@@ -213,7 +230,7 @@ const CATEGORY_DEFINITIONS = [
       /assertions?\s+match\s+very\s+long/i,
       /test\s+can\s+hang/i,
       /brittle\s+to\s+minor\s+copy\s+edits/i,
-      /doesn['’]t\s+actually\s+assert/i,
+      /doesn['']t\s+actually\s+assert/i,
       /claims?\s+.*\s+but/i,
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/github/audit-copilot-comments.mjs:234]: All 6 findings fixed in 7e828e1:

1. **curly-apostrophe patterns**: Restored U+2019 matching in incomplete_coverage and misleading_tests regex — changed from /doesn['']t/ back to /doesn['’]t/ (straight + curly right quote).
2. **--sleep-ms without --checkpoint-file**: Sleep now applies between fetches unconditionally, not only when checkpoint-file is set.
3. **checkpoint parent directory**: saveCheckpoint now creates parent directories via mkdir(recursive: true) before writing.
4. **gate-evidence prompt**: Rewrote 'runs on draft gate by default' to 'opt-in for the draft gate (uncomment to add)'.
5. **--checkpoint-file whitespace**: Added non-empty validation (empty after trim → parseError).
6. Added corresponding test for whitespace-only --checkpoint-file rejection.

825 tests pass (scripts), full verify green.

File: scripts/github/audit-copilot-comments.mjs [hunk 3]
```diff
@@ -737,9 +860,51 @@ export function parseAuditCopilotCommentsCliArgs(argv) {
...
+  }
+
+  // ── Fresh fetch path ─────────────────────────────────────────────
+  if (comments === null) {
+    comments = await fetchAllReviewComments(options.repo, { env, ghCommand, retryMax, retryBaseMs });
+
+    if (checkpointFile) {
+      await saveCheckpoint(checkpointFile, { stage: "after-comments", repo: options.repo, comments });
+      await sleepStep(sleepMs);
+    }
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/github/audit-copilot-comments.mjs:908]: All 6 findings fixed in 7e828e1:

1. **curly-apostrophe patterns**: Restored U+2019 matching in incomplete_coverage and misleading_tests regex — changed from /doesn['']t/ back to /doesn['’]t/ (straight + curly right quote).
2. **--sleep-ms without --checkpoint-file**: Sleep now applies between fetches unconditionally, not only when checkpoint-file is set.
3. **checkpoint parent directory**: saveCheckpoint now creates parent directories via mkdir(recursive: true) before writing.
4. **gate-evidence prompt**: Rewrote 'runs on draft gate by default' to 'opt-in for the draft gate (uncomment to add)'.
5. **--checkpoint-file whitespace**: Added non-empty validation (empty after trim → parseError).
6. Added corresponding test for whitespace-only --checkpoint-file rejection.

825 tests pass (scripts), full verify green.

File: scripts/github/audit-copilot-comments.mjs [hunk 4]
```diff
@@ -359,6 +376,89 @@ const RECOMMENDATION_DEFINITIONS = [
...
+    return null;
+  } catch {
+    return null;
+  }
+}
+
+async function saveCheckpoint(checkpointPath, data) {
+  await writeFile(checkpointPath, `${JSON.stringify(data)}\n`, "utf8");
+}
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/github/audit-copilot-comments.mjs:451]: All 6 findings fixed in 7e828e1:

1. **curly-apostrophe patterns**: Restored U+2019 matching in incomplete_coverage and misleading_tests regex — changed from /doesn['']t/ back to /doesn['’]t/ (straight + curly right quote).
2. **--sleep-ms without --checkpoint-file**: Sleep now applies between fetches unconditionally, not only when checkpoint-file is set.
3. **checkpoint parent directory**: saveCheckpoint now creates parent directories via mkdir(recursive: true) before writing.
4. **gate-evidence prompt**: Rewrote 'runs on draft gate by default' to 'opt-in for the draft gate (uncomment to add)'.
5. **--checkpoint-file whitespace**: Added non-empty validation (empty after trim → parseError).
6. Added corresponding test for whitespace-only --checkpoint-file rejection.

825 tests pass (scripts), full verify green.

File: .pi/dev-loop/defaults.yaml [hunk 5]
```diff
@@ -241,3 +257,82 @@ personas:
...
+      - Check that the draft gate review comment exists and references the
+        correct head SHA.
+      - Check that the pre-approval gate review comment exists when the PR
+        is in a ready/merge state.
+      - Flag any PR that transitions to ready/merge states without visible
+        gate evidence for the current head.
+      - When draft-first enforcement is configured, verify that a draft-gate
+        comment was posted before the PR was marked ready.
+      - This persona runs on draft gate by default; on first PR it checks for
+        any gate evidence (not only draft_gate since no prior gate exists).
```
[the remark this answers is not in the document]
Comment (mfittko) [reply .pi/dev-loop/defaults.yaml]: All 6 findings fixed in 7e828e1:

1. **curly-apostrophe patterns**: Restored U+2019 matching in incomplete_coverage and misleading_tests regex — changed from /doesn['']t/ back to /doesn['’]t/ (straight + curly right quote).
2. **--sleep-ms without --checkpoint-file**: Sleep now applies between fetches unconditionally, not only when checkpoint-file is set.
3. **checkpoint parent directory**: saveCheckpoint now creates parent directories via mkdir(recursive: true) before writing.
4. **gate-evidence prompt**: Rewrote 'runs on draft gate by default' to 'opt-in for the draft gate (uncomment to add)'.
5. **--checkpoint-file whitespace**: Added non-empty validation (empty after trim → parseError).
6. Added corresponding test for whitespace-only --checkpoint-file rejection.

825 tests pass (scripts), full verify green.

File: scripts/github/audit-copilot-comments.mjs [hunk 6]
```diff
@@ -717,6 +816,26 @@ export function parseAuditCopilotCommentsCliArgs(argv) {
...
+        throw parseError("--sleep-ms must be a non-negative integer");
+      }
+      options.sleepMs = parsed;
+      continue;
+    }
+
+    if (token === "--checkpoint-file") {
+      options.checkpointFile = requireOptionValue(args, "--checkpoint-file", parseError).trim();
+      continue;
+    }
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/github/audit-copilot-comments.mjs:833]: All 6 findings fixed in 7e828e1:

1. **curly-apostrophe patterns**: Restored U+2019 matching in incomplete_coverage and misleading_tests regex — changed from /doesn['']t/ back to /doesn['’]t/ (straight + curly right quote).
2. **--sleep-ms without --checkpoint-file**: Sleep now applies between fetches unconditionally, not only when checkpoint-file is set.
3. **checkpoint parent directory**: saveCheckpoint now creates parent directories via mkdir(recursive: true) before writing.
4. **gate-evidence prompt**: Rewrote 'runs on draft gate by default' to 'opt-in for the draft gate (uncomment to add)'.
5. **--checkpoint-file whitespace**: Added non-empty validation (empty after trim → parseError).
6. Added corresponding test for whitespace-only --checkpoint-file rejection.

825 tests pass (scripts), full verify green.

File: scripts/github/audit-copilot-comments.mjs [hunk 7]
```diff
@@ -737,9 +866,52 @@ export function parseAuditCopilotCommentsCliArgs(argv) {
...
+        prs = checkpoint.prs;
+      } else if (checkpoint.stage === "after-comments") {
+        await sleepStep(sleepMs);
+        prs = await fetchAllPullRequests(options.repo, { env, ghCommand, retryMax, retryBaseMs });
+        await saveCheckpoint(checkpointFile, { stage: "after-prs", repo: options.repo, comments, prs });
+      }
+    }
+
+    // If checkpoint was corrupted/absent, fall through to fresh fetch.
+  }
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/github/audit-copilot-comments.mjs:900]: All 3 findings fixed in dbc5e6e:

1. **Partial checkpoint fallback**: Resume path now checks Array.isArray(checkpoint.comments) and Array.isArray(checkpoint.prs) for after-prs stage. Unrecognized stages or missing prs → fall through to fresh fetch.
2. **Redundant dynamic import**: Removed dynamic import of mkdir in saveCheckpoint; uses existing top-level import.
3. **runGhJsonWithRetry defaults**: Added default ghCommand='gh' and env=process.env in parameter destructuring for safe exported API usage.

Added corresponding tests (runGhJsonWithRetry default ghCommand, partial checkpoint with missing prs). 827 tests pass.

File: scripts/github/audit-copilot-comments.mjs [hunk 8]
```diff
@@ -359,6 +376,91 @@ const RECOMMENDATION_DEFINITIONS = [
...
+  } catch {
+    return null;
+  }
+}
+
+async function saveCheckpoint(checkpointPath, data) {
+  const { mkdir } = await import("node:fs/promises");
+  await mkdir(path.dirname(checkpointPath), { recursive: true });
+  await writeFile(checkpointPath, `${JSON.stringify(data)}\n`, "utf8");
+}
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/github/audit-copilot-comments.mjs:451]: All 3 findings fixed in dbc5e6e:

1. **Partial checkpoint fallback**: Resume path now checks Array.isArray(checkpoint.comments) and Array.isArray(checkpoint.prs) for after-prs stage. Unrecognized stages or missing prs → fall through to fresh fetch.
2. **Redundant dynamic import**: Removed dynamic import of mkdir in saveCheckpoint; uses existing top-level import.
3. **runGhJsonWithRetry defaults**: Added default ghCommand='gh' and env=process.env in parameter destructuring for safe exported API usage.

Added corresponding tests (runGhJsonWithRetry default ghCommand, partial checkpoint with missing prs). 827 tests pass.

File: scripts/github/audit-copilot-comments.mjs [hunk 9]
```diff
@@ -359,6 +376,91 @@ const RECOMMENDATION_DEFINITIONS = [
...
+
+function isAuthError(stderrText) {
+  return /\bHTTP 401\b/i.test(stderrText);
+}
+
+/**
+ * Retry a gh JSON call with exponential backoff on 403/429.
+ * Fails immediately on 401 (auth) or other non-retryable errors.
+ */
+export async function runGhJsonWithRetry(args, { env, ghCommand, retryMax = DEFAULT_RETRY_MAX, retryBaseMs = DEFAULT_RETRY_BASE_MS } = {}) {
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/github/audit-copilot-comments.mjs]: All 3 findings fixed in dbc5e6e:

1. **Partial checkpoint fallback**: Resume path now checks Array.isArray(checkpoint.comments) and Array.isArray(checkpoint.prs) for after-prs stage. Unrecognized stages or missing prs → fall through to fresh fetch.
2. **Redundant dynamic import**: Removed dynamic import of mkdir in saveCheckpoint; uses existing top-level import.
3. **runGhJsonWithRetry defaults**: Added default ghCommand='gh' and env=process.env in parameter destructuring for safe exported API usage.

Added corresponding tests (runGhJsonWithRetry default ghCommand, partial checkpoint with missing prs). 827 tests pass.

**Comment by mfittko** [comment]
### Gate review: `pre_approval_gate`

**Reviewed head SHA:** `dbc5e6ee952e62d0f008eb3c549d1ead6633d9ec`
**Verdict:** clean

**Findings summary:** ci: 2 Copilot review rounds addressed, round 3 clean (0 findings). 827 tests pass. CI green.

**Next action:** Merge PR.
